NumLib/ODESolver/ODETypes.h

```diff
@@ -0,0 +1,210 @@
+/**
+ * \copyright
+ * Copyright (c) 2012-2016, OpenGeoSys Community (http://www.opengeosys.org)
+ *            Distributed under a Modified BSD License.
+ *              See accompanying file LICENSE.txt or
+ *              http://www.opengeosys.org/project/license
+ *
+ */
+
+#ifndef NUMLIB_ODETYPES_H
+#define NUMLIB_ODETYPES_H
+
+// TODO: move that file somewhere else
+
+#include <initializer_list>
+#include <cassert>
+
+
+// Dense Eigen matrix/vector //////////////////////////////////////////
+// always enabled
+
+#include <Eigen/LU>
+
+
+inline void setMatrix(Eigen::MatrixXd& m,
+                      Eigen::MatrixXd::Index const rows, Eigen::MatrixXd::Index const cols,
+                      std::initializer_list<double> values)
+{
+    using IndexType = Eigen::MatrixXd::Index;
+    assert((IndexType) values.size() == rows*cols);
+
+    auto it = values.begin();
+    for (IndexType r=0; r<rows; ++r) {
+        for (IndexType c=0; c<cols; ++c) {
+            m(r, c) = *(it++);
+        }
+    }
+}
+
+inline void setMatrix(Eigen::MatrixXd& m, Eigen::MatrixXd const& tmp)
+{
+    m = tmp;
+}
+
+inline void addToMatrix(Eigen::MatrixXd& m,
+                        Eigen::MatrixXd::Index const rows, Eigen::MatrixXd::Index const cols,
+                        std::initializer_list<double> values)
+{
+    using IndexType = Eigen::MatrixXd::Index;
+    assert((IndexType) values.size() == rows*cols);
+
+    auto it = values.begin();
+    for (IndexType r=0; r<rows; ++r) {
+        for (IndexType c=0; c<cols; ++c) {
+            m(r, c) += *(it++);
+        }
+    }
+}
+
+namespace NumLib
+{
+inline void oneShotLinearSolve(Eigen::MatrixXd& A, Eigen::VectorXd& rhs, Eigen::VectorXd& x)
+{
+    Eigen::FullPivLU<Eigen::MatrixXd> slv;
+    slv.compute(A);
+    x = slv.solve(rhs);
+}
+
+inline double norm(Eigen::VectorXd const& x) { return x.norm(); }
+
+}
+
+inline void setVector(Eigen::VectorXd& v, std::initializer_list<double> values)
+{
+    assert((std::size_t) v.size() == values.size());
+    auto it = values.begin();
+    for (std::size_t i=0; i<values.size(); ++i) v[i] = *(it++);
+}
+
+
+#ifdef USE_PETSC
+
+// Global PETScMatrix/PETScVector //////////////////////////////////////////
+
+#include "MathLib/LinAlg/PETSc/PETScMatrix.h"
+#include "MathLib/LinAlg/PETSc/PETScVector.h"
+
+
+namespace NumLib
+{
+inline void oneShotLinearSolve(MathLib::PETScMatrix& A, MathLib::PETScVector& rhs,
+                               MathLib::PETScVector& x)
+{
+    (void) A; (void) rhs; (void) x;
+    // TODO implement
```

Comment:
Probably I will drop the `oneShotLinearSolve()` functions in the follow-up PR that integrates the ODE lib into the rest of OGS6.